lib/Sema/CSDiag.cpp

```diff
@@ -5410,13 +5410,75 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
 
     return true;
   }
-  
-  // If we found an exact match, this must be a problem with a conversion from
-  // the result of the call to the expected type.  Diagnose this as a
-  // conversion failure.
-  if (calleeInfo.closeness == CC_ExactMatch)
+
+  // We have found exact much for the arguments, which means this is either
+  // generic or conversion failure, let's try to diagnose that.
+  if (calleeInfo.closeness == CC_ExactMatch) {
+    // This is not an initializer, we don't know what might be wrong.
+    if (!isa<TypeExpr>(fnExpr))
+      return false;
+
+    // We have an initializer expression for some type (might be generic),
+    // before giving up trying to diagnose this situation, let's double
+    // check if we have any unresolved parameters, and try to diagnose that
+    // situation. This might happen in the situation like
+    // this - struct S<A, B> { init(a: A) {} } and call S(a: 1),
+    // initializer is going to be (A) -> S<A, B>, A is going to be inferred from
+    // initializer argument and B going to be unspecified.
+    auto &candidate = calleeInfo.candidates[0];
+
+    auto resultType = candidate.getResultType();
+    // We can only try to diagnose bound generic types.
+    if (!resultType->is<BoundGenericType>())
+      return false;
+
+    auto BGT = resultType->castTo<BoundGenericType>();
+    llvm::SmallSet<ArchetypeType *, 2> missingParams;
+    for (auto &genericParam : BGT->getGenericArgs()) {
+      if (auto archetype = genericParam->getAs<ArchetypeType>()) {
+        // First, let's consider all of the generic arguments missing.
+        missingParams.insert(archetype);
+      }
+    }
+
+    auto argumentType = candidate.getArgumentType();
+    // Argument type is not a tuple, might be a structural problem.
+    if (!argumentType->is<TupleType>())
+      return false;
+
+    for (auto arg : argumentType->castTo<TupleType>()->getElements()) {
```

Comment:
Argument mismatches are going to be diagnosed before this diagnostic is considered because it's only triggered when there is an exact match on the application, I'll add a test for [T] case and see if this works with correct code or not. Thanks!